File: app/src/main/java/com/buggzy/smsrestroom/MainActivity.java
package com.buggzy.smsrestroom;

import android.Manifest;
import android.content.Context;
import android.content.DialogInterface;
import android.content.IntentFilter;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.provider.Telephony;
import android.support.annotation.NonNull;
import android.support.v4.app.ActivityCompat;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.text.InputType;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.EditText;
import android.widget.TextView;

public class MainActivity extends AppCompatActivity {

    private static final int SMS_PERMISSION_CODE = 0;
    private static String RESTUrl = "";
    String rest_url_key = "rest_url";
    private boolean isRunning = false;
    private SmsBroadcastReceiver smsBroadcastReceiver;

    public static String getRESTUrl() {
        return RESTUrl;
    }

    public void setRESTUrl(String url) {
        RESTUrl = url;
        saveSettings();
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        loadSettings();

        refreshStatusText();
    }

    private void loadSettings() {
        Context context = getApplicationContext();
        SharedPreferences sharedPref = context.getSharedPreferences(
            getString(R.string.preference_file_key), Context.MODE_PRIVATE);

        RESTUrl = sharedPref.getString(rest_url_key, "");
    }

    private void saveSettings() {
        Context context = getApplicationContext();
        SharedPreferences sharedPref = context.getSharedPreferences(
            getString(R.string.preference_file_key), Context.MODE_PRIVATE);

        SharedPreferences.Editor editor = sharedPref.edit();
        editor.putString(rest_url_key, RESTUrl);
        editor.apply();
    }

    private void startListener() {
        if (!isRunning) {
            smsBroadcastReceiver = new SmsBroadcastReceiver();
            registerReceiver(smsBroadcastReceiver, new IntentFilter(Telephony.Sms.Intents.SMS_RECEIVED_ACTION));
        }
        isRunning = true;
    }

    private void stopListener() {
        if (isRunning) {
            unregisterReceiver(smsBroadcastReceiver);
        }
        isRunning = false;
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        refreshStatusText();
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
    }

    protected void refreshStatusText() {
        String statusText = "";

        statusText = statusText + "Status: " +
            (isRunning ? "Running" : "Stopped") + "\n";

        statusText = statusText + "SMS permissions: " +
            (hasReadSmsPermission() ? "OK" : "failed") + "\n";

        statusText = statusText + "REST URL: " +
            (getRESTUrl().equals("") ? "<undefined>" : getRESTUrl()) + "\n";

        TextView stControl = findViewById(R.id.statusText);
        stControl.setText(statusText);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        int id = item.getItemId();

        //noinspection SimplifiableIfStatement
        if (id == R.id.action_settings) {
            promptForURL();
//            new RestNotifier().execute("123", "123");
            return true;
        }

        if (id == R.id.action_permissions) {
            requestSmsPermissions();
            return true;
        }

        if (id == R.id.action_start) {
            startListener();
            refreshStatusText();
            return true;
        }

        if (id == R.id.action_stop) {
            stopListener();
            refreshStatusText();
            return true;
        }

        return super.onOptionsItemSelected(item);
    }

    private boolean hasReadSmsPermission() {
        return ContextCompat.checkSelfPermission(MainActivity.this,
            Manifest.permission.READ_SMS) == PackageManager.PERMISSION_GRANTED &&
            ContextCompat.checkSelfPermission(MainActivity.this,
                Manifest.permission.RECEIVE_SMS) == PackageManager.PERMISSION_GRANTED;
    }

    private void requestSmsPermissions() {
        ActivityCompat.requestPermissions(MainActivity.this, new String[]{Manifest.permission.READ_SMS, Manifest.permission.RECEIVE_SMS, Manifest.permission.INTERNET},
            SMS_PERMISSION_CODE);
    }

    private void promptForURL() {
        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        builder.setTitle("REST URL");

        final EditText input = new EditText(this);
        input.setInputType(InputType.TYPE_CLASS_TEXT);
        input.setText(getRESTUrl());
        builder.setView(input);

        builder.setPositiveButton("OK", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                setRESTUrl(input.getText().toString());
                refreshStatusText();
            }
        });
        builder.setNegativeButton("Cancel", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {
                dialog.cancel();
            }
        });

        builder.show();
    }
}

File: app/src/main/java/com/buggzy/smsrestroom/MainActivity.kt
package com.buggzy.smsrestroom

import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import androidx.appcompat.app.AlertDialog
import com.buggzy.smsrestroom.base.BaseActivity
import com.buggzy.smsrestroom.extensions.*
import com.buggzy.smsrestroom.receivers.ServiceReceiver
import com.karumi.dexter.Dexter
import com.karumi.dexter.MultiplePermissionsReport
import com.karumi.dexter.PermissionToken
import com.karumi.dexter.listener.PermissionRequest
import com.karumi.dexter.listener.multi.MultiplePermissionsListener
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.android.synthetic.main.content_main.*
import kotlinx.android.synthetic.main.dialog_url.*
import org.jetbrains.anko.activityManager
import org.jetbrains.anko.stopService

class MainActivity : BaseActivity(), MultiplePermissionsListener {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        setSupportActionBar(toolbar)
        if (Preferences.isRunning) {
            if (!activityManager.isServiceRunning<MainService>()) {
                startForegroundService<MainService>()
            }
        } else {
            if (activityManager.isServiceRunning<MainService>()) {
                stopService<MainService>()
            }
        }
        refreshStatusText()
        Dexter.withActivity(this)
            .withPermissions(*allAppPermissions)
            .withListener(this)
            .check()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            R.id.action_settings -> {
                promptForURL()
                true
            }
            R.id.action_start -> {
                if (!Preferences.isRunning) {
                    Preferences.isRunning = true
                    refreshStatusText()
                    restartForegroundService<MainService>()
                }
                true
            }
            R.id.action_stop -> {
                if (Preferences.isRunning) {
                    Preferences.isRunning = false
                    refreshStatusText()
                    cancelAlarm<ServiceReceiver>()
                    if (activityManager.isServiceRunning<MainService>()) {
                        stopService<MainService>()
                    }
                }
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }

    private fun promptForURL() {
        val dialog = AlertDialog.Builder(this).apply {
            setTitle("Base URL")
            setView(View.inflate(applicationContext, R.layout.dialog_url, null))
            setNegativeButton(getString(android.R.string.cancel), null)
            setPositiveButton(getString(android.R.string.ok), null)
        }.create()
        dialog.setOnShowListener {
            val input = dialog.input_url
            input.setText(Preferences.baseUrl)
            input.setSelection(input.text.length)
            dialog.getButton(AlertDialog.BUTTON_POSITIVE).setOnClickListener {
                Preferences.baseUrl = input.text.toString().trim()
                refreshStatusText()
                dialog.cancel()
            }
        }
        dialog.show()
    }

    private fun refreshStatusText() {
        statusText.text = StringBuilder().apply {
            append("Android ID: $androidId\n")
            append("Status: ${if (Preferences.isRunning) "running" else "stopped"}\n")
            append("Base URL: ${Preferences.baseUrl}\n")
        }.toString()
    }

    override fun onPermissionsChecked(report: MultiplePermissionsReport?) {}

    override fun onPermissionRationaleShouldBeShown(permissions: MutableList<PermissionRequest>?, token: PermissionToken?) {}
}
